### PR TITLE
Customize launching connector to support cwd and enviroment variable

### PR DESCRIPTION
**Pull request by testforstephen**

### CODE REVIEW

File: com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/DebugUtility.java [hunk 1]
```diff
@@ -91,6 +91,68 @@ public static IDebugSession launch(VirtualMachineManager vmManager, String mainC
         return new DebugSession(vm);
     }
 
+    public static IDebugSession launch(VirtualMachineManager vmManager, String mainClass, String programArguments, String vmArguments, List<String> classPaths,
```
Comment (andxu) [inline com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/DebugUtility.java]: Delete the old launch method.

File: com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/adapter/handler/LaunchRequestHandler.java [hunk 2]
```diff
@@ -91,11 +94,30 @@ public void handle(Command command, Arguments arguments, Response response, IDeb
...
         sourceProvider.initialize(options);
 
+        // Append environment to native environment.
+        String[] envp = null;
+        if (launchArguments.env != null && !launchArguments.env.isEmpty()) {
+            Map<String, String> environment = new HashMap<>(System.getenv());
+            for (Entry<String, String> entry : launchArguments.env.entrySet()) {
+                // For duplicated variable, show a warning message.
+                if (environment.containsKey(entry.getKey())) {
+                    logger.warning(String.format("There is the same variable '%s' existed in the system environment, "
```
Comment (akaroml) [inline com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/adapter/handler/LaunchRequestHandler.java]: Nice!

File: com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/adapter/handler/LaunchRequestHandler.java [hunk 3]
```diff
@@ -91,11 +94,30 @@ public void handle(Command command, Arguments arguments, Response response, IDeb
...
+                            + "the program will use the newly configured variable to override the system environment variable.", entry.getKey()));
```
Comment (akaroml) [inline com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/adapter/handler/LaunchRequestHandler.java]: There are duplicated environment variables. The values specified in launch.json will be used. Here are the duplicated entries: env1, env2.

File: com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/adapter/handler/LaunchRequestHandler.java [hunk 4]
```diff
@@ -91,11 +94,30 @@ public void handle(Command command, Arguments arguments, Response response, IDeb
         }
         sourceProvider.initialize(options);
 
+        // Append environment to native environment.
+        String[] envp = null;
+        if (launchArguments.env != null && !launchArguments.env.isEmpty()) {
+            Map<String, String> environment = new HashMap<>(System.getenv());
+            for (Entry<String, String> entry : launchArguments.env.entrySet()) {
+                // For duplicated variable, show a warning message.
+                if (environment.containsKey(entry.getKey())) {
```
Comment (akaroml) [inline com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/adapter/handler/LaunchRequestHandler.java]: Suggest finding all the duplicated entries and output one aggregated warning.

File: com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/adapter/handler/LaunchRequestHandler.java [hunk 5]
```diff
@@ -91,11 +94,30 @@ public void handle(Command command, Arguments arguments, Response response, IDeb
         }
         sourceProvider.initialize(options);
 
+        // Append environment to native environment.
+        String[] envp = null;
```
Comment (akaroml) [inline com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/adapter/handler/LaunchRequestHandler.java]: envp -> envVars

File: com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/VirtualMachineLauncher.java [hunk 6]
```diff
@@ -0,0 +1,423 @@
...
+import com.sun.jdi.VirtualMachineManager;
+import com.sun.jdi.VoidValue;
+import com.sun.jdi.connect.Connector;
+import com.sun.jdi.connect.IllegalConnectorArgumentsException;
+import com.sun.jdi.connect.ListeningConnector;
+import com.sun.jdi.connect.VMStartException;
+import com.sun.jdi.event.EventQueue;
+import com.sun.jdi.request.EventRequestManager;
+
+public class VirtualMachineLauncher {
```
Comment (akaroml) [inline com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/VirtualMachineLauncher.java]: This implementation is a bit overwhelming. I'll continue the review later.

File: com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/DebugUtility.java [hunk 7]
```diff
@@ -242,4 +257,57 @@ public static void resumeThread(ThreadReference thread) {
...
+     * <p>source argument:
+     * <pre>["path=C:\\ProgramFiles\\java\\bin", "JAVA_HOME=C:\\ProgramFiles\\java"]</pre>
+     *
+     * <p>after encoded:
+     * <pre>"29\npath=C:\\ProgramFiles\\java\\bin30\nJAVA_HOME=C:\\ProgramFiles\\java"</pre>
+     *
+     * @param argument the string array arguments
+     * @return the encoded string
+     */
+    public static String encodeArrayArgument(String[] argument) {
```
Comment (akaroml) [inline com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/DebugUtility.java]: We already referenced the apache commons lib. We can still keep the function but please use  org.apache.commons.lang3.StringUtils.join as implementation

File: com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/DebugUtility.java [hunk 8]
```diff
@@ -242,4 +257,57 @@ public static void resumeThread(ThreadReference thread) {
...
+    /**
+     * Decode the encoded string to the original string array.
+     *
+     * @param argument the decoded string
+     * @return the original string array argument
+     * @throws IllegalArgumentException
+     *          when the argument is not encoded by the rules defined in the encodeArrayArgument.
+     */
+    public static String[] decodeArrayArgument(String argument) throws IllegalArgumentException {
+        if (argument == null) {
```
Comment (akaroml) [inline com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/DebugUtility.java]: Here we can use org.apache.commons.lang3.StringUtils.split as impelemntation.

File: com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/AdvancedLaunchingConnector.java [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+        } catch (IOException | IllegalConnectorArgumentsException e) {
+            process.destroy();
+            throw new VMStartException(String.format("VM did not connect within given time: %d ms", ACCEPT_TIMEOUT), process);
+        }
+
+        vm.setLaunchedProcess(process);
+        return vm;
+    }
+
+    private static String[] constructLaunchCommand(Map<String, ? extends Argument> launchingOptions, String address) {
```
Comment (akaroml) [inline com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/AdvancedLaunchingConnector.java]: Can we use the implementation from the base class?

File: com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/AdvancedLaunchingConnector.java [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+    }
+
+    /**
+     * Parses the given command line into separate arguments that can be passed
+     * to <code>Runtime.getRuntime().exec(cmdArray)</code>.
+     *
+     * @param args command line as a single string.
+     * @return the arguments array.
+     */
+    private static List<String> parseArguments(String cmdStr) {
```
Comment (akaroml) [inline com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/AdvancedLaunchingConnector.java]: Can we use the implementation from the base class?

File: com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/AdvancedLaunchingConnector.java [hunk 11]
```diff
@@ -0,0 +1,157 @@
...
+        // 2. a quote " followed by whatever, until another quote "
+        Matcher m = Pattern.compile("([^\"]\\S*|\".+?\")\\s*").matcher(cmdStr);
+        while (m.find()) {
+            String arg = m.group(1).replaceAll("^\"|\"$", ""); // Remove surrounding quotes.
+            list.add(arg);
+        }
+        return list;
+    }
+
+    class AdvancedStringArgumentImpl extends StringArgumentImpl implements StringArgument {
```
Comment (akaroml) [inline com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/AdvancedLaunchingConnector.java]: Why do we need to implement this one? Seems the implementation does not do much...

Comment (testforstephen) [reply com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/AdvancedLaunchingConnector.java]: The original StringArgumentImpl doesn't expose it's constructor method. Cannot new a StringArgumentImpl in our connector.

File: com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/DebugUtility.java (hunk 7 above)
Comment (testforstephen) [reply com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/DebugUtility.java]: My implementation is not to join the array but use the special rule to encode the string array. Can you explain why need to use `StringUtils.join` as implementation?

File: com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/AdvancedLaunchingConnector.java (hunk 9 above)
Comment (testforstephen) [reply com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/AdvancedLaunchingConnector.java]: fixed.

File: com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/AdvancedLaunchingConnector.java (hunk 10 above)
Comment (testforstephen) [reply com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/AdvancedLaunchingConnector.java]: fixed.

File: com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/DebugUtility.java (hunk 7 above)
Comment (akaroml) [reply com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/DebugUtility.java]: We can pair two calls to achieve what we want: StringUtils.join(args, '\n') and String.Utils.split(str, '\n')

Comment (testforstephen) [reply com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/DebugUtility.java]: if the environment variable may contain "\n" by itself, the join/split cannot handle this case.

Comment (testforstephen) [reply com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/DebugUtility.java]: Use URLEncoder/Decoder to encode/decode string array.

**Pull request merged**